compiler/sigmatch.nim

```diff
@@ -1279,7 +1282,10 @@ proc typeRel(c: var TCandidate, f, aOrig: PType,
         result = isGeneric
       else:
         result = typeRel(c, ff, aa, flags)
-
+      
+      if fRange.kind == tyArray and fRange.indexType.kind == tyGenericParam:
```

Comment:
`fRange.kind == tyArray` doesn't make sense, I think the lines above are wrong:

```nim
if prev == nil:
  put(c, fRange, a.indexType)
  fRange = a
```

`fRange = a` should be `fRange = a.indexType`, then we can just check if `fRange.kind == tyGenericParam`.

Reply:
I'm of the opinion that the whole `tyArray` branch of typeRel needs to be revisited, not just this part. I didn't try to mess with what was already there. Sorry this PR is so patchy. It would take me a while to understand how the statics are packed into the node structure.

Reply:
And to be clear, I did look at that briefly and `fRange = a` makes zero sense to me at all, so if you have an idea about that you are further along then me.

Reply:
wait, something just clicked for me. `fRange` is being set to `a` because if it has no binding it's assumed that it can accept `a`, so that is the substitution. What if the generic parameter has constraints though? I think that might be another error. Just because the range is a generic type variable doesn't mean it accepts any type. That's an unconstrained generic type variable.